### ISSUE TITLE
Bug: Moving the mouse over a huge container mostly filled with invisible sprites results in a big performance hit

### ISSUE THREAD

**Issue by TobiasWehrum** [pixijs/pixijs#9237]
### Current Behavior

In an scene with a huge container with most sprites being `visible=false;`, moving the mouse over the canvas results in a big performance hit.

### Expected Behavior

Invisible objects have no impact while moving the mouse over the canvas. (Pixi v6 had no performance penality.)

### Steps to Reproduce

1. Open https://jsfiddle.net/TobiasW/7qf6har0/32/
2. Observe the framerate being e.g. around 30 FPS
3. Move your mouse over the canvas.
4. While continually moving the mouse, observe the framerate drop to e.g. around 6 FPS.

### Environment

- **`pixi.js` version**: 7.2.0-rc
- **Browser & Version**: Firefox 110.0.1 (64-Bit)
- **OS & Version**: Windows 10
- **Running Example**: https://jsfiddle.net/TobiasW/7qf6har0/32/

### Additional Information

This seems to be similiar to #9156, but with `visible = false;` being ignored instead of `interactive = false; interactiveChildren = false; hitArea = { contains: () => false }`.

**Comment by rwampler** [comment]
I was just creating almost exactly the example, which matches my case too, thanks!

One slight difference I had is that I do want some interaction events (clicks) on visible sprites, but don't need anything from movement

Example: https://jsfiddle.net/jwoq4m2z/

**Comment by bigtimebuddy** [comment]
Thanks for the great examples. Definitely still seems like an issue.

**Comment by bigtimebuddy** [comment]
No performance issues with `sprites.interactiveChildren = false`, which makes me think it's the cost of the large children list in `sprites`.
https://jsfiddle.net/bigtimebuddy/7uw0oh9a/ 

### Additional Context

We made a somewhat fundamental behavior change in v7 when we tried to align the new events with the DOM. For instance, in the DOM, if you have an element overlapping another element, it _blocks_ interaction. Pixi v7 now does this too. We also added bubbling which was another requested feature from DOM events.

The downside of this approach is that everything is hit-tested by default, however, you can now use `eventMode = 'none'` is somewhat equivalent to `pointer-events: none` in CSS. This will completely ignore the interactivity and allow pass-thru interactions.

The performance issue with 300K Sprites is that it's _very expensive_ to loop through all these elements _even if_ we are setting `eventMode` to `none` OR ignoring hidden elements (although this is a good optimization we should consider adding that would benefit deeply nested scenes). Also, the default `eventMode` is `auto` (formerly `interaction=false`), but this can be changed to `eventMode: 'none'` in the Application or Renderer constructor arguments to disable all interaction (we are considering changing this default in v8).

Here's some more context https://github.com/pixijs/pixijs/pull/9089#issuecomment-1398113742 about `eventMode`.

**Comment by bigtimebuddy** [comment]
@Zyie I think there are a couple ideas how we could marginally improve perf with events:

* Add visibility filter to the `hitPruneFn` and `all` functions in `EventBoundary`
* Consider an opt-out (or opt-in) boolean for disabling global move events if not needed

**Comment by TobiasWehrum** [comment]
> The performance issue with 300K Sprites is that it's very expensive to loop through all these elements [...]

I tried that: https://jsfiddle.net/TobiasW/q51hstvr/10 (click to toggle the "count invisible objects) loop on/off)

- If I'm running through the that loop 1 times, I see no framerate difference with/without the loop.
- If I'm moving the mouse over the window, the framerate drops from 30-40 FPS to 6-7 FPS.

So I think at least on my machine, ignoring hidden elements would not be a marginal performance improvement, but quite a massive one.

**Comment by TobiasWehrum** [comment]
> The performance issue with 300K Sprites is that it's very expensive to loop through all these elements [...]

I tried that: https://jsfiddle.net/TobiasW/q51hstvr/35 (click to toggle the "count invisible objects) loop on/off)

- If I'm running through the that loop 1 times, I see no framerate difference with/without the loop.
- If I'm moving the mouse over the window, the framerate drops from 30-40 FPS to 6-7 FPS.

So I think at least on my machine, ignoring hidden elements would not be a marginal performance improvement, but quite a massive one.

**Comment by bigtimebuddy** [comment]
Try this: https://jsfiddle.net/bigtimebuddy/L2y19zqn/ does it make any difference?

**Comment by TobiasWehrum** [comment]
Sadly none whatsoever (tested with Firefox).

---

I also made a curious discovery - with my original script (https://jsfiddle.net/TobiasW/7qf6har0/32) and PixiJS 7.2.0-rc I can reproduce the problem in Firefox, but not in Chrome.

- Firefox
 - 6.x: No framedrop
 - 7.x: Framedrop 30 -> 6 FPS
 - 7.2.0-rc: Framedrop 30 -> 6 FPS

- Chrome
 - 6.x: No framedrop
 - 7.x: Framedrop 60 -> 10 FPS
 - 7.2.0-rc: No framedrop

---

And one more bonus while we're at things that are weird:
- Chrome, PixiJS 7.x, with Devtools Console open: Framedrop 60 -> 0 (it freezes as long as I move the mouse, and immediately unfreezes when I stop)

**Comment by Zyie** [comment]
Hey all

I have a PR up to try and fix these issues here: #9246 

Before: https://jsfiddle.net/TobiasW/7qf6har0/32
After: https://jsfiddle.net/kv46xsjr/

It seems to fix this example

@rwampler For your use case i have add options to disable `move`, `globalMove`, `click`, and `wheel` events that allows you to turn on/off large parts on the event system

```
new Application({
    events: {
        move: false,
        globalMove: false,
        click: true,
        wheel: false
    }
})
```

**Comment by TobiasWehrum** [comment]
@Zyie 

Tested in Firefox on Windows 10:
- Before: FPS drop 35 FPS -> 6 FPS
- After: FPS drop 37 FPS -> 17 FPS

So, it got much better, but that still seems pretty harsh.

I'm assuming you are doing something like checking if `visible === false` and skip all of those? I wonder why it still has such a performance penality, when I didn't see any when trying to loop over all objects in the container in https://github.com/pixijs/pixijs/issues/9237#issuecomment-1458601121

**Comment by Zyie** [comment]
> * Before: FPS drop 35 FPS -> 6 FPS
> * After: FPS drop 37 FPS -> 17 FPS
> 
> So, it got much better, but that still seems pretty harsh.
> 

oh i forgot to test this in firefox

It's really weird why the performance is so bad compared to chrome. I need to dig into this some more.
I did do some quick testing and found that the performance is roughly the same as using pixi v6, is that the same for you?

**Comment by TobiasWehrum** [comment]
In Firefox with v6 (https://jsfiddle.net/TobiasW/phuxr24v/2/), the drop is less pronounced for me: 34 FPS -> 24 FPS.

Yeah, the Chrome thing is really weird. In 7.x, both browsers show a similar framedrop, but with 7.2.0-rc (so even before your fix), the framedrop seems gone in Chrome (but is still there in Firefox).

**Comment by rwampler** [comment]
Awesome, tested out 7.2.0-rc.2 with the new flags and performance for me back to or better than pre-v7. Even without flags, performance seems to much better.

I am seeing new weird behavior around wrong sprites receiving events, but guessing that's a bug in my sprite caching and re-use logic, and maybe exposed now by the consolidated interaction loop and different order of events. Will debug and raise new issue if able to reproduce.

Thanks for all your help!

**Comment by bigtimebuddy** [comment]
I'm going to mark this as fixed with #9246 (available in 7.2.0-rc.2)

**Comment by rwampler** [comment]
Working on example, but with some debugging it does look like that sprite event mismatch might be related to custom hit areas and recent changes: https://github.com/pixijs/pixijs/issues/9257
